pkg/controllers/management/clusterupstreamrefresher/cluster_upstream_refresher.go

```diff
@@ -0,0 +1,323 @@
+		cluster, err = c.migrateEksRefreshCronSetting(cluster)
+		if err != nil {
+			return cluster, err
+		}
+		cluster.Annotations[eksRefreshCronMigrated] = "true"
+		if cluster, err = c.clusterClient.Update(cluster); err != nil {
+			return cluster, err
+		}
+	}
+
+	providerRefreshInterval, err := getProviderRefreshInterval(provider)
+	if err != nil {
+		return cluster, err
+	}
+
+	nextRefreshTime, err := getNextProviderRefreshTime(providerRefreshInterval, lastRefreshTime)
+	if err != nil {
+		return cluster, err
+	}
+
+	if !time.Now().After(nextRefreshTime) {
+		return cluster, err
+	}
+
+	cluster, err = c.refreshClusterUpstreamSpec(cluster, provider)
+	if err != nil {
+		return cluster, err
+	}
+
+	c.clusterEnqueueAfter(key, providerRefreshInterval)
+
+	return cluster, nil
+}
+
+// getProviderAndReadyStatus returns the managed cluster provider of the given
+// cluster and whether it is ready to be refresh and synced.
+func getProviderAndReadyStatus(cluster *mgmtv3.Cluster) (string, bool) {
+	// for other cloud drivers, please edit HERE
+	switch {
+	case cluster.Spec.EKSConfig != nil:
+		if cluster.Status.EKSStatus.UpstreamSpec == nil {
+			logrus.Infof("initial upstream spec for cluster [%s] has not been set by cluster handler yet, skipping", cluster.Name)
+			return apimgmtv3.ClusterDriverEKS, false
+		}
+		return apimgmtv3.ClusterDriverEKS, true
+	default:
+		return noKEv2Provider, false
+	}
+}
+
+// getProviderRefreshInterval returns the duration that should pass between
+// refreshing a cluster created by the given cloud provider.
+func getProviderRefreshInterval(provider string) (time.Duration, error) {
+	providerRefreshSetting, err := settings.GetSettingByID(strings.ToLower(fmt.Sprintf(refreshCronSettingFormat, provider)))
+	if err != nil {
+		return 0, err
+	}
+
+	refreshInterval := providerRefreshSetting.GetInt()
+
+	return time.Duration(refreshInterval) * time.Second, nil
+}
+
+// getNextProviderRefreshTIme returns the next time a KEv2 cluster is due to refresh given
+// its last refresh time and cloud provider.
+func getNextProviderRefreshTime(refreshInterval time.Duration, lastRefreshTime string) (time.Time, error) {
+	if lastRefreshTime == "" {
+		return time.Now(), nil
+	}
+
+	lastRefreshUnix, err := strconv.ParseInt(lastRefreshTime, 10, 64)
+	if err != nil {
+		return time.Time{}, fmt.Errorf("unable to parse last KEv2 refresh time [%s]: %v", lastRefreshTime, err)
+	}
+
+	return time.Unix(lastRefreshUnix, 0).Add(refreshInterval), nil
+}
+
+// migrateEksRefreshCronSetting migrate deprecated EKS refresh cron setting to new format
+// and remove deprecated eks-refresh-cron setting object
+func (c *clusterRefreshController) migrateEksRefreshCronSetting(cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
+	eksCronSetting, err := c.settingsCache.Get(eksRefreshCronDeprecated)
+	if err != nil {
+		return cluster, err
+	}
+
+	eksSchedule, err := cron.Parse(eksCronSetting.Default)
```

Comment:
Should use value not default or you are not getting the users setting. But see this conversation as it will change this code: https://github.com/rancher/rancher/pull/31771#discussion_r599707655

Reply:
In situations where you have access tot he setting like this, you can just use <setting>.Get() if you need the effective value.

Reply:
If` Value` was set for `eks-cron-refresh` setting then it is taken, if not then `Default` is used.